docs/madr/decisions/003-logging.md

```diff
@@ -0,0 +1,245 @@
+# Logging standard
+
+* Status: accepted
+
+## Context and Problem Statement
+
+Seamless troubleshooting is a feature. We should keep more attention to the logging in Kuma.
+
+## Considered Options
+
+* Adopting consistent logging standard.
+
+## Decision Outcome
+
+Chosen option: "Adopting consistent logging standard", because it's the only option.
+
+### Positive Consequences
+
+* Better UX for users.
+* Better troubleshooting means both users and maintainers spent less time debugging an issue.
+
+### Negative Consequences
+
+* More effort to try to follow the guidelines.
+
+## Pros and Cons of the Options
+
+### Adopting consistent logging standard
+
+#### Convention
+
+Ideally, all those points are linted.
+
+1) Logs are lowercase. Can be uppercase when you start with proper noun.
+_GOOD_
+```go
+log.Info("something has happened")
+log.Info("Envoy is starting")
+```
+_BAD_
+```go
+log.Info("Something has happened")
+```
+
+2) Do not include `.` at the end of the sentence if this is the only sentence in the log
+_GOOD_
+```go
+log.Info("something has happened")
+log.Info("something has happened. Restarting the process")
+```
+_BAD_
+```go
+log.Info("something has happened.")
+```
+
+3) Either log or return an error. Don't do both, unless you want to log more details on DEBUG level.
+_GOOD_
+```go
+if err != nil {
+	return errors.Wrap(err, "more info")
+}
+...
+if err != nil {
+	log.Error(err, "more info")
+	return nil
+}
+...
+if err != nil {
+	log.V(1).Info("much more details about the error")
+    return errors.Wrap(err, "more info")
+}
+```
+_BAD_
+```go
+if err != nil {
+    log.Error(err, "more info")
+	return errors.Wrap(err, "more info")
+}
+```
+
+4) Use structured logging instead of `fmt.Sprintf`
+_GOOD_
+```go
+log.Info("dp connected", "name", dpName)
+```
+_BAD_
+```go
+log.Info(fmt.Sprintf("dp %s connected", dpName))
+```
+
+5) Carry the same context to multiple logs using WithValues
+_GOOD_
+```go
+log := logger.WithValues("proxy", dpName, "mesh", mesh)
+log.Info("connected")
+log.Info("reconciled")
+log.Info("disconnected")
+```
+_BAD_
+```go
+logger.Info("connected", proxy", dpName, "mesh", mesh)
+logger.Info("reconciled", proxy", dpName, "mesh", mesh)
+logger.Info("disconnected", proxy", dpName, "mesh", mesh)
+```
+
+6) Be mindful about logger name. Try to not repeat the information.
+_GOOD_
+```go
+var log = core.Log.WithName("xds-server")
+log.Info("starting")
+```
+_BAD_
+```go
+var log = core.Log.WithName("xds-server")
+log.Info("starting XDS server")
+```
+
+7) Don't use ellipsis
+_GOOD_
+```go
+log.Info("starting")
+```
+_BAD_
+```go
+log.Info("starting...")
+```
+
+8) Log before executing the action
```

Comment:
Is there guidance about what "an action" is? Or is it just a "you know it when you see it" thing.